### ISSUE TITLE
AdductomicsR package

### ISSUE THREAD

**Issue by JosieLHayes** [Bioconductor/Contributions#920]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/JosieLHayes/adductomicsR
- Associated data repository (too large to put on Github): https://berkeley.box.com/s/titfjzlwhqmmjjwm9pw9uqv3ahpzotxm

Confirm the following by editing each check box to '[x]'

- [ x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x ] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x ] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [ x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x ] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x ] The 'devel' branch for new packages and features.
- [x ] The stable 'release' branch, made available every six
      months, for bug fixes.
- [ x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
For your large data resource it is probably appropriate to make this an 'experiment hub' package, where the large data is hosted in ExperimentHub, and a light-weight package provides access to this resource. I'll move this issue to the moderation queue, but a first step will be to work with @lshep to get the data available through ExperimentHub. See the vignette 'Creating an ExperimentHub package' at https://bioconductor.org/packages/devel/ExperimentHub.

**Comment by JosieLHayes** [comment]
I have now uploaded the data package, adductData to Github https://github.com/JosieLHayes/adductData, and the raw data to ExperimentHub S3 (adductData). @lshep please can you check the metadata files as discussed. Thanks!

**Comment by LiNk-NY** [comment]
Lori, @lshep, is everything okay on the data side ? 

Usually maintainers submit data packages first and then the software packages
so that the dependencies are in order. How should this issue proceed? Does Josie need to
create a new issue or just reference the data package in a comment?

@JosieLHayes Please refer to https://github.com/Bioconductor/Contributions after further instructions. 

Regards,
Marcel

**Comment by JosieLHayes** [comment]
Thank you. I'll add that to the issue and make sure it's working before debugging adductomicsR.

On Nov 5, 2018 4:21 AM, lshep <notifications@github.com> wrote:

You can add the package to this issue by commenting AdditionalPackages: https://github.com/JosieLHayes/adductData as described here<https://github.com/Bioconductor/Contributions#submitting-related-packages>. Its okay that the data package was submitted second just please have that working before trying to debug the software package. Don't forget to add the webhook to both the data package and the software package so that both will trigger a new build on version bump.

**Comment by lshep** [comment]
Sorry those ERRORs are on out end.  I am in the process of updating the SPB to R 3.6 and until this is complete there will be ERRORs.

**Comment by JosieLHayes** [comment]
Ok, thanks for letting me know.

**Comment by JosieLHayes** [comment]
Is there any news on when R 3.6 will be available? Also when it comes online do I have to bump the version?

Josie

**Comment by lshep** [comment]
R 3.6 has been updated for MAC and Windows but we are having some issues with the linux install.  Since I know you have been waiting on this I will kick off manual builds of your packages now even without linux  -  after this please version bump for your own corrections.   When we get the linux R 3.6 up I will also do a manual run of your packages.   Cheers and sorry for any inconvenience this delay has caused.

**Comment by JosieLHayes** [comment]
Great, thanks Lori. Much appreciated!
I'll start debugging.

**Comment by JosieLHayes** [comment]
I'm getting "ABNORMAL" for build results. Any ideas why? I did change the version to be 0.99.1 on both packages - could it be this?

**Comment by federicomarini** [comment]
Hi @JosieLHayes , I think it is due to the Single Package Builder being under maintenance. 
I am getting the same result in my recently submitted package - see https://github.com/Bioconductor/Contributions/issues/937.

See also this: https://stat.ethz.ch/pipermail/bioc-devel/2018-November/014354.html

**Comment by lshep** [comment]
I think I have fixed the issue -  please let us know if you continue to experience issues.

**Comment by federicomarini** [comment]
Excellent! Do we just need to bump the version to have the build triggered?

**Comment by lshep** [comment]
I manually triggered a new build for you -  hopefully when you push yourself it will be ok

**Comment by federicomarini** [comment]
Great, thanks. I'll follow up on the other issue to keep this more focused on `adductomicsR` itself

**Comment by JosieLHayes** [comment]
Thanks Lori and Federico for working on this!

**Comment by JosieLHayes** [comment]
I am having some trouble accessing the data from experimentHub for my software packages examples. For the vignette, which works I access the data using:

eh  = ExperimentHub()
temp = query(eh, 'adductData')
temp[['EH1957']]  #first mzXML file
file.rename(file.path(hubCache(temp),'1957'), file.path(hubCache(temp), 'ORB35017.mzXML'))
Then hubCache(temp) is the filepath I need for my examples with ORB35017.mzXML being on of the mzXML examples.

Does this have to be done for every example in the software package?

Thanks,

Josie

**Comment by JosieLHayes** [comment]
Hi,
I'm getting the following error during R CMD check in the build report:
Error in library.dynam(lib, package, package.lib) : 
  shared object mzR.so not found

Is it possible to reinstall mzR on R 3.6? I think this may be the problem.
Thanks

**Comment by JosieLHayes** [comment]
Hi Lori @lshep 
Is there any way I can speed up the ExperimentHub download further? I keep getting a warning in R CMD check that it takes longer than 5 minutes. I have only put the download in once and as its the same session the files are still there for the other examples, however it is still too long.
Thanks, 
Josie

**Comment by lshep** [comment]
No but perhaps instead of using the entire object you can use a subset for examples?

**Comment by lshep** [comment]
Also your reviewer @LiNk-NY  Could answer any non hub related questions regarding any ERROR/WARNINGS/TIMEOUTS

**Comment by lshep** [comment]
While @LiNk-NY will review your package -  please note -  it looks like you write files to the users home directory -  this should not be done, there should always be an option of where to save data and in examples/vignettes temporary directory should be used.  The following were written to the packagebuilders home directory 
```
dductQuantif_peakList_2018-11-16.csv
adductQuantif_peakList_2018-11-29.csv
adductQuantif_peakList_2018-12-07.csv
adductQuantif_peakList_2018-12-08.csv
adductQuantif_peakList_2018-12-10.csv
adductQuantif_peakList_2018-12-11.csv
```

**Comment by JosieLHayes** [comment]
Thanks Lori, I’ll change that.

**Comment by JosieLHayes** [comment]
I have been trying to reduce the length of time it takes to do R CMD check, but when I remove examples or set them to dontrun{} I then get an error that I do not have enough runnable examples. Please could you advise me a way around this?

Thanks
Josie

**Comment by JosieLHayes** [comment]
Hi @lshep and @mtmorgan,
I am getting abnormal for some of the builds. Please can you look into this for me.
Thanks

**Comment by LiNk-NY** [comment]
Hi Josie, @JosieLHayes 
It seems like there was work being done on some of the platforms and the package
wasn't built on those. 
Lori can elaborate further. 

Thanks, 
MR

**Comment by JosieLHayes** [comment]
Ok thanks for letting me know. I reran it with a version bump and it was different so I'll wait till later and version bump again.

**Comment by JosieLHayes** [comment]
Hi @LiNk-NY 
The package runs with no errors or warnings on two of the systems but has not been run on the third. Have updates to the system completed yet? I just wanted to check whether there is anything more you require me to do before you can review it.
Thanks,
Josie

**Comment by LiNk-NY** [comment]
Hi Josie, @JosieLHayes 
Thanks for staying on top of it! 
I will have a look at the packages soon.
Best,
Marcel

**Comment by JosieLHayes** [comment]
Thank you!
Best,
Josie

**Comment by LiNk-NY** [comment]
Hi Josie, 

Thanks for your submission to Bioconductor. 
Please see the review below. 
If you have any questions, post them here. 

Best regards,
Marcel 

----

# adductomicsR #920

## DESCRIPTION

* Looks good.
* No need to use `LazyData` or `LazyLoad` here. It only applies to data in the
`data/` folder.
* All your imports don't need to have a minimum version requirement but you
can leave it like this.
* Add a `BugReports` field for filing GitHub issues.

## NAMESPACE

* Looks good !

## vignettes/

* You don't need to mention the version of R required as this will change with
Bioconductor release and devel versions. The current submission will go into
Bioc-devel
* Enclose code within text with backticks '``'
* `eval=TRUE` is the chunk default, there's no need to repeat it
* Consider formatting the code in the code chunks for legibility
(e.g., in 'Quantify adducts')

## R/

* Check if you can use related and established Bioconductor classes to represent
your data
* Class names usually start with a capital letter.
* Create class accessor functions to avoid the use of `@`.
* Format code for readability (see `c` function)
* `nCores` argument(s) should be conservative by default rather than using all
available cores
* `flush.console()` not really needed
* Avoid changing user options, change them back to what they were on exiting
the function
* Use `message` instead of `cat` for showing status messages
* Refrain from saving plots on the user's behalf, users should save plots
themselves in the appropriate format from the graphics device. At the very
least, a `NULL` output dir should not save any files rather than in the
current directory
* It's not clear how you're using the `dotProdSimClust` function inside
`minDotProdSpec` and returning the output of the inner function as the return
value of the outer function (see `R/dotProdSpectra.R`)
* There are a lot of hard-coded values in `IsotopicDistributionMod` perhaps
these values can go in a table?
* Use `&&` for robustly testing single conditions on either side of `&&`,
`&` returns a logical vector where possible
* Check curly braces in `specSimPepId`, it seems like there is an extra closing
brace

Minor:
* `if (!ncol(x))` is equivalent to `if (ncol(x) == 0)`
* No need to test for `TRUE` as `hclustRt == TRUE` just use `if (hclustRt)`,
in the case that the output is not certain to be logical, use `isTRUE()`
* Use the inverse of logicals instead of testing for a `FALSE` value (e.g.,
use `!duplicated(targTable$peptide)` R/adductQuant.R:L119)
* Try `seq_along` instead of `seq_len(length(x))`
* No need to include `x` in `elements = list(x, ...)` since it is already in
the signature

**Comment by JosieLHayes** [comment]
Thanks Marcel. I’ll get to work on these points.

**Comment by JosieLHayes** [comment]
Ok, thanks for letting me know - I am easily confused!!
I also wanted to ask - is it ok to put localHub=T in the examples once it is in the cache to decrease the R CMD check time?

**Comment by lshep** [comment]
The localHub=TRUE will not decrease the check time -  its an option for if there is no internet connection but requires that the file already have been downloaded to the cache.  I would avoid using this option in code.


Lori Shepherd

Bioconductor Core Team

Roswell Park Cancer Institute

Department of Biostatistics & Bioinformatics

Elm & Carlton Streets

Buffalo, New York 14263

________________________________

**Comment by JosieLHayes** [comment]
Ok, good to know!

**Comment by JosieLHayes** [comment]
Hi @LiNk-NY 

I have addressed all your comments from the review. Below are some notes on each point. Please can you let me know what you think.
Thank you.

**Review commments**
Check if you can use related and established Bioconductor classes to represent your data - There are not many classes for mzXML files and so I had to make new ones in this instance. I did check again just in case this has changed though.
Class names usually start with a capital letter - both classes now start with a capital letter.

Create class accessor functions to avoid the use of @. I have added class accessor functions as suggested.

Format code for readability (see c function). All code has been formatted.
nCores argument(s) should be conservative by default rather than using all available cores- the default now is nCores=NULL which runs on one core.
flush.console() not really needed - removed.
Avoid changing user options, change them back to what they were on exiting the function. I have changed stringsAsFactors=F in adductSpecGen back to =T on exit.
Use message instead of cat for showing status messages - all changed.
Refrain from saving plots on the user's behalf, users should save plots themselves in the appropriate format from the graphics device. At the very least, a NULL output dir should not save any files (instead of saving in the current directory) - now there is an outputPlotDir parameter. When this is NULL no plots are saved.
It's not clear how you're using the dotProdSimClust function inside minDotProdSpec and returning the output of the inner function as the return value of the outer function (see R/dotProdSpectra.R) - I have put a better explanation in dotprodspectra.R params.
There are a lot of hard-coded values in IsotopicDistributionMod perhaps these values can go in a table? I implemented this but it seemed to take longer and so I removed it and kept the hard coded version.
Use && for robustly testing single conditions on either side of &&, & returns a logical vector where possible -I have changed this where appropriate.
Check curly braces in specSimPepId, it seems like there is an extra closing brace. I have removed this.

**Minor**
if (!ncol(x)) is equivalent to if (ncol(x) == 0) - I have not changed this as it throws an error when I do 
No need to test for TRUE as hclustRt == TRUE just use if (hclustRt) in the case that the output is not certain to be logical, use isTRUE() - I have changed these.
Use the inverse of logicals instead of testing for a FALSE value (e.g., 
use !duplicated(targTable$peptide) R/adductQuant.R:L119)- I have altered this where I think appropriate.
Try seq_along instead of seq_len(length(x)) - I have changed all these.
No need to include x in elements = list(x, ...) since it is already in the signature - I have removed x in all these cases.

**Comment by JosieLHayes** [comment]
Hi @LiNk-NY,

Just wondering whether you have had a chance to do a second review of the adductomicsR package yet?

Thanks,
Josie

**Comment by JosieLHayes** [comment]
Fantastic, thanks Marcel. I will make those changes once it has been added
to the devel repo.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/JosieLHayes.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("adductomicsR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/adductomicsR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/JosieLHayes.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("adductData")`. The package 'landing page' will be created at

https://bioconductor.org/packages/adductData

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by JosieLHayes** [comment]
Thank you!

I have the ssh key set up and will be in touch if I have further questions.
